### ISSUE TITLE
Need to compile Sass with catalog start

### ISSUE THREAD

**Issue by mattfelten** [interactivethings/catalog#314]
I'm looking at integrating Catalog for the documentation end of a design system I'm building. It's Sass files for CSS and React components packaging the HTML. I have the React components working in the page right now (Awesome!), but I'm not sure how to get the CSS to build and included (hopefully globally). 

I've included the file main Sass file into the Catalog component but it's not actually compiling anything. It's expecting a CSS file.

Any advice?

**Comment by piotrf** [comment]
@mattfelten, AFAIK it should be as straight-forward as requiring your `.scss` file with all its imports in `styleguide.jsx`:

```
import React from 'react';
import ReactDOM from 'react-dom';
import { configureRoutes } from 'catalog';

import Styles from "my-styles.scss";

const routes = configureRoutes({
  title: 'My Styleguide',
  basePath: '/',
  pages: [
    { component: require('./docs/intro'), path: '/', title: 'Introduction' },
  ]
});
```

You can also contextualise your SASS files and require them directly from component React js or jsx file:

```
import React from 'react';

require('./PrimaryButton.scss');

const PrimaryButton = props => <button className={`my-button`}>{props.children}</button>;
```

**Comment by herrstucki** [comment]
There are two ways:

- You can pre-build your SASS code and include the bilt CSS, see https://github.com/facebookincubator/create-react-app/blob/master/packages/react-scripts/template/README.md#adding-a-css-preprocessor-sass-less-etc
- You can use a `catalog.config.js` file to add sass-loader, see https://docs.catalog.style/guides/webpack-babel

Let me know if that works.

**Comment by mattfelten** [comment]
I got the pre-built Sass version to work so I'm going with that for now. Thanks!

So I like the second option better because it seems cleaner, but I couldn't figure out how to get it to work. If you have time, here's the 3 different ways I tried: https://gist.github.com/mattfelten/854685ef6cca02b177901176398e38dd I don't know enough about webpack I don't think.

**Comment by herrstucki** [comment]
I *think* you need to _prepend_ the sass-loader to `module.rules` but I haven't tested it. Otherwise your first option looks like the best one (the other two definitely don't work).

**Comment by nickjohnson-dev** [comment]
I realize this issue is a bit old, and I am happy starting a new issue if you prefer it @herrstucki, but I wanted to check if it's possible to get a snippet of a catalog.config.js file with a working sass-loader setup.

I've tried a few different ways to merge my webpack config in, using a working /\.scss$/ loader rule from another project, but always run into various errors.

Also, is it possible to import sass *only* for certain specimens, to avoid affecting the layout of the overall Catalog?

Let me know if there's a better place to ask these questions, and I'll do that. 😅 

Thanks!

**Comment by herrstucki** [comment]
@nickjohnson-dev Doesn't the CRA guide work for you (i.e. prebuilding SCSS and then importing the CSS in Catalog)?

Off the top of my head, something like this may work:

```js
// catalog.config.js
module.exports = {
  webpack: (webpackConfig) => {
    webpackConfig.module.rules = [
      {test: /\.scss$/, use: 'sass-loader', options: {}},
      ...webpackConfig.module.rules
    ])
    return webpackConfig;
  }
}
```

**Comment by nickjohnson-dev** [comment]
@herrstucki I'm not rigidly opposed to just compiling my SCSS code and then importing it, but it seemed like you had gone to the trouble to support custom webpack loaders, and I didn't see an obvious reason why this wasn't working.

Using the code you provided, with the loaders necessary to load in the styles:

```javascript
module.exports = {
  webpack: config => {
    config.module.rules = [
      {
        test: /\.scss$/,
        use: 'style-loader!css-loader!sass-loader',
      },
      ...config.module.rules,
    ];
    return config;
  },
};
```

...I am getting this error:

```bash
This relative module was not found:

* ./styles.scss in ./catalog/index.js
```

I'm using the same versions of all the relevant dependencies that I use in other projects, and I'm unsure what the difference here would be.

**Comment by herrstucki** [comment]
It seems like your styles.scss is not in the correct place? The error message it should be at `catalog/styles.scss` …

**Comment by nickjohnson-dev** [comment]
Yeah, I did make sure that the files were in the right place and everything, and I was still getting the error no matter what I changed with the location of the file.

Fortunately, for my use case I think that I've gotten a good workflow down with just compiling the SCSS files in my npm "prestart" script and then loading them as css, so I'm not blocked by this issue.

I appreciate you taking a look at it though!

**Comment by alanbentley** [comment]
Having the same issue when trying to add a new module rule to handle the scss - no matter where I place the sccs file I get "This relative module was not found".

I've opted to use a postinstall script to compile the scss - but Ideally I'd like to get it all working from the webpack config - I'm not sure whats effecting the path.

**Comment by metakermit** [comment]
I also ended up using quite a bit of time trying to load .scss files directly, but it seems there are currently some performance and other issues with Webpack plugins for Sass. I ended up following [the CRA instructions](https://github.com/facebook/create-react-app/blob/master/packages/react-scripts/template/README.md#adding-a-css-preprocessor-sass-less-etc) to the letter and got a solution I'm quite happy with:

```
// package.json
{
  "name": "catalog",
  "dependencies": {
    "catalog": "^3.0.0-rc.4",
    "node-sass-chokidar": "^1.1.0",
    "npm": "^5.7.1",
    "npm-run-all": "^4.1.2",
    "react": "^16.2.0",
    "react-dom": "^16.2.0"
  },
  "scripts": {
    "build-css":
      "node-sass-chokidar --include-path ./catalog --include-path ./node_modules catalog/ -o catalog/",
    "watch-css":
      "npm run build-css && node-sass-chokidar --include-path ./catalog --include-path ./node_modules catalog/ -o catalog/ --watch --recursive",
    "catalog-start-js": "catalog start",
    "catalog-build-js": "catalog build",
    "catalog-start": "npm-run-all -p watch-css catalog-start-js",
    "catalog-build": "npm-run-all build-css catalog-build-js",
    "start": "npm run catalog-start",
    "build": "npm run catalog-build"
  }
}
```

Then I can place an *index.scss* in e.g. *components/core/index.scss* and load it from index.js as:

```
import React from "react";
import ReactDOM from "react-dom";
import { Catalog, pageLoader } from "catalog";
import "./components/core/index.css";

const pages = [
  {
    path: "/",
    title: "Introduction",
    content: pageLoader(() => import("./WELCOME.md"))
  }
];

ReactDOM.render(
  <Catalog title="MyCatalog" pages={pages} />,
  document.getElementById("catalog")
);
```

I also ignore .css files in *.gitignore*:

```
catalog/**/*.css
```

Also worth noting – for some reason whatever method of trying to import the styles as [per the Catalog docs](https://docs.catalog.style/configuration/imports), it doesn't work. That's why I ended up just importing the css using ES6 modules. For example this didn't load the style:

```
const pages = [
  {
    styles: ["./components/core/index.css"],
    path: "/",
    title: "Introduction",
    content: pageLoader(() => import("./WELCOME.md"))
  }

**Comment by metakermit** [comment]
OK, my config got a bit more complicated because I started using [lerna](https://lernajs.io/) for managing multiple npm packages inside a single repo. Ended up overriding my *catalog.config.js* to get Sass loading to work from the root of the project:

```javascript
const path = require("path");

const modulesPath = path.resolve(__dirname);

module.exports = {
  webpack: config => {
    config.module.rules.push({
      test: /\.scss$/,
      loaders: [
        "style-loader",
        "css-loader",
        {
          loader: "sass-loader",
          options: {
            includePaths: [modulesPath]
          }
        }
      ],
      include: modulesPath
    });
    return config;
  }
};
```

More info on the setup [here](https://github.com/lerna/lerna/issues/1367#issuecomment-380827147).
